File: Rust/src/gui.rs
use std::io::{self, Write};

use crate::apple::Apple;
use crate::node::Node;
use crate::snake::Snake;
use crate::terminal::goto_xy;

pub struct GUI {
    pub map: [[u8; 22]; 22],
}

impl GUI {
    pub fn new() -> Self {
        GUI { map: [[0; 22]; 22] }
    }

    fn place_borders(&mut self) {
        for i in 0..22 {
            self.map[0][i] = b'#';
            self.map[21][i] = b'#';
            self.map[i][0] = b'#';
            self.map[i][21] = b'#';
        }
    }

    fn place_snake(&mut self, snake: &Snake) {
        let head: &Box<Node> = &snake.head;
        self.map[(head.x + 1) as usize][(head.y + 1) as usize] = b'X';

        let mut curr: Option<&crate::node::Node> = head.next.as_deref();

        while let Some(node) = curr {
            let x: usize = (node.x + 1) as usize;
            let y: usize = (node.y + 1) as usize;

            if x < 22 && y < 22 {
                self.map[x][y] = b'O';
            }
            curr = node.next.as_deref();
        }
    }

    fn place_apple(&mut self, apple: &Apple) {
        let x: usize = (apple.x + 1) as usize;
        let y: usize = (apple.y + 1) as usize;

        if x < 22 && y < 22 {
            self.map[x][y] = b'@';
        }
    }

    pub fn gen_gui(&mut self, snake: &Snake, apple: &Apple, score: i32) {
        goto_xy(1, 1);

        self.map = [[b' '; 22]; 22];

        self.place_borders();
        self.place_snake(snake);
        self.place_apple(apple);

        for j in 0..22 {
            for i in 0..22 {
                let field: u8 = self.map[i][j];

                if field == b'#' {
                    print!("##");
                } else {
                    print!("{} ", field as char);
                }
            }

            if j == 0 {
                print!("  Score:  {}", score);
            }

            println!();
        }

        io::stdout().flush().unwrap();
    }

    pub fn game_over() {
        goto_xy(1, 23);

        println!("         ___                 ");
        println!("        / __|__ _ _ __  ___  ");
        println!("       | (_ / _` | ''  \\/ -_) ");
        println!("        \\___\\__,_|_|_|_\\___| ");
        println!("         ___                 ");
        println!("        / _ \\_ _____ _ _     ");
        println!("       | (_) \\ V / -_) ''_|   ");
        println!("        \\___/ \\_/\\___|_|     ");
    }
}

File: lua/gui.lua
local terminal = require("terminal")

local gui = {}
gui.__index = gui

function gui.new()
    local g = {map = {}}
    setmetatable(g, gui)

    for y = 1, 22 do
        g.map[y] = {}
        for x = 1, 22 do
            g.map[y][x] = " "
        end
    end

    return g
end

function gui:place_borders()
    for i = 1, 22, 1 do
        self.map[1][i] = '#'
        self.map[22][i] = '#'
        self.map[i][1] = '#'
        self.map[i][22] = '#'
    end
end

function gui:place_snake(snake)
    if snake.head == nil then
        return
    end

    local curr = snake.head

    self.map[curr.y + 2][curr.x + 2] = 'X'

    curr = curr.next

    while curr ~= nil do
        self.map[curr.y + 2][curr.x + 2] = 'O'
        curr = curr.next
    end
end

function gui:place_apple(apple)
    self.map[apple.y + 2][apple.x + 2] = '@'
end

function gui:draw(snake, apple, score)
    terminal.goto_xy(1, 1)

    for y = 1, 22 do
        for x = 1, 22 do
            self.map[y][x] = " "
        end
    end

    self:place_borders()
    self:place_snake(snake)
    self:place_apple(apple)

    local out = ""

    for j = 1, 22, 1 do
        for i = 1, 22, 1 do
            local field = self.map[j][i]

            if field == '#' then
                out = out .. "##"
            else
                out = out .. field .. " "
            end
        end

        if j == 1 then
            out = out .. "  Score:  " .. score
        end

        out = out .. "\n"
    end

    io.write(out)
end

function gui:game_over()
    terminal.goto_xy(1, 23)

    io.write("         ___                 \n")
    io.write("        / __|__ _ _ __  ___  \n")
    io.write("       | (_ / _` | ''  \\/ -_) \n")
    io.write("        \\___\\__,_|_|_|_\\___| \n")
    io.write("         ___                 \n")
    io.write("        / _ \\_ _____ _ _     \n")
    io.write("       | (_) \\ V / -_) ''_|   \n")
    io.write("        \\___/ \\_/\\___|_|     \n")
end

return gui